src/components/document/editable-document-content.tsx

```diff
@@ -1,40 +1,40 @@
-import React, { useContext, useRef } from "react";
+import React, { useContext, useRef, useState } from "react";
 import classNames from "classnames";
-import { AppConfigContext, IAppConfigContext } from "../../app-config-context";
+import { AppConfigContext } from "../../app-config-context";
 import { CanvasComponent } from "./canvas";
 import { DocumentContextReact } from "./document-context";
 import { FourUpComponent } from "../four-up";
 import { useDocumentContext } from "../../hooks/use-document-context";
 import { useGroupsStore, useUIStore, useUserStore } from "../../hooks/use-stores";
-import { ToolbarComponent, ToolbarConfig } from "../toolbar";
+import { ToolbarComponent } from "../toolbar";
 import { EditableToolApiInterfaceRef, EditableToolApiInterfaceRefContext } from "../tools/tool-api";
 import { DocumentModelType } from "../../models/document/document";
 import { ProblemDocument } from "../../models/document/document-types";
 import { WorkspaceMode } from "../../models/stores/workspace";
-import { getToolContentInfoByTool } from "../../models/tools/tool-content-info";
-import { IToolButtonConfig } from "../tool-button";
+import { clone } from "mobx-state-tree";
+import { ToolbarModelType } from "../../models/stores/app-config-model";
 
 import "./editable-document-content.scss";
 
 interface IToolbarProps {
   document: DocumentModelType;
-  toolbar?: ToolbarConfig;
-}
-
-function getToolbarIcon(tool: IToolButtonConfig, appConfig: IAppConfigContext) {
-  if (tool.isTileTool) {
-    return getToolContentInfoByTool(tool.name).Icon;
-  } else {
-    return appConfig.appIcons?.[tool.iconId];
-  }
+  toolbar: ToolbarModelType;
 }
 
 const DocumentToolbar: React.FC<IToolbarProps> = ({ toolbar, ...others }) => {
   const appConfig = useContext(AppConfigContext);
-  const toolbarConfig = toolbar?.map(tool => ({ icon: getToolbarIcon(tool, appConfig), ...tool }));
-  return toolbarConfig
-          ? <ToolbarComponent key="toolbar" config={toolbarConfig} {...others} />
-          : null;
+
+  // The toolbar prop represents the app's configuration of the toolbar
+  // It is cloned here in the document so changes to one document's toolbar
+  // do not effect another document's toolbar.
+  // Currently the toolbar model is not modified, but it seems safer to do this.
+  // The cloned model is stored in state so it isn't recreated on each render
+  const [toolbarModel] = useState<ToolbarModelType>(() => {
+      // The new model is passed the appIcons as its environment, so the model
+      // can lookup an app level Icon if needed.
+      return clone(toolbar, { appIcons: appConfig.appIcons });
+  });
```

Comment:
```suggestion
  // The new model is passed the appIcons as its environment, so the model
  // can look up an app level Icon if needed.
  const toolbarModelRef = useRef(clone(toolbar, { appIcons: appConfig.appIcons }));
```
For a persistent instance that is never changed, `useRef` seems more appropriate than `useState`.

Reply:
I thought about this a lot and settled on useState.

Here are some reasons why:

1. lazy initialization: 
The problem with the basic `useRef` approach is that it will clone the toolbar on every render. Only the first clone will be used (future clones will get thrown away), but it seems wasteful to do that clone each time.  With useState the cone is in a function so it is only run once.

There is a pattern for doing lazy `useRef` like this:https://reactjs.org/docs/hooks-faq.html#:~:text=%20avoid%20re-creating%20the%20useref
But that seemed more complicated than just using `useState`.

2. This really seems like state of the component. It isn't changing, but its whole purpose is to be passed to a child component during render. It seems to me the main goal of useRef is to support changing or accessing an object outside of the render loop.

3. I didn't verify this, but I think it will show up in the react dev tools so it would be easier to inspect.

Reply:
>lazy initialization:
The problem with the basic useRef approach is that it will clone the toolbar on every render. Only the first clone will be used (future clones will get thrown away), but it seems wasteful to do that clone each time. With useState the clone is in a function so it is only run once.

Good point -- I should have thought about that.

>There is a pattern for doing lazy useRef like this:https://reactjs.org/docs/hooks-faq.html#:~:text=%20avoid%20re-creating%20the%20useref
But that seemed more complicated than just using useState.

Agreed, but I would just do something like
```typescript
  const toolbarModelRef = useRef<ToolbarModelType>(null);
  if (!toolbarModelRef.current) {
    toolbarModelRef.current = clone(toolbar, { appIcons: appConfig.appIcons });
  }
```
I don't see the value in the intervening function in their example.

>This really seems like state of the component. It isn't changing, but its whole purpose is to be passed to a child component during render. It seems to me the main goal of useRef is to support changing or accessing an object outside of the render loop.

I don't think that characterization of `useRef` is correct. According to the docs: "The useRef() Hook isn’t just for DOM refs. The “ref” object is a generic container whose current property is mutable and can hold any value, similar to an instance property on a class."

To me, the goal of `useState` is to represent something that when changed will trigger a re-render. For values that aren't changed and/or should not trigger a re-render `useRef` seems more natural.

All of that said, we're splitting hairs here. I would use a ref, but your argument for state has some merit as well so if my counter-argument doesn't convince you feel free to stick with what you have.

Reply:
It is splitting hairs but sometimes I enjoy figuring this stuff out. (not always). 😄 

I also think it kind of establishes a pattern of how MST models should be managed by their root components. At least in the case where they are passed into children components as properties. 

So my counter argument is that we would actually want to re-render the child component if for some reason the toolbarModel changes. This is because it is passed as a property.  If the property was assigned to some other model it would make sense to completely re-render the child.

In the current code this isn't possible because there is no `setToolbarModel`, but I could see some developer wanting to change this in the future. If a ref is used they might not remember they need to do something additional to trigger the re-render.

So I guess the rule of thumb I'm thinking of is: 

> If a value is passed to a child as a property then it should be state or a property of the parent. This way changes trigger re-renders.
>
> With the following exception:
> When value change frequently, child components are expensive to re-render, and the child component explicitly supports the pattern of passing a ref. 

I'm assuming there isn't a penalty for using state.  

I also realize I'm trying to make rules without having enough React hook experience. So I'm sure I'm oversimplifying.  I hope I have enough time to try to apply this pattern to more code so I can see how well it works.

Reply:
>It is splitting hairs but sometimes I enjoy figuring this stuff out. (not always).

Agreed -- these sorts of theoretical discussions can be fun and they're a good way to learn.

>So my counter argument...

I think that's a counter-counter-argument at this point. 😉 

>...is that we would actually want to re-render the child component if for some reason the toolbarModel changes. This is because it is passed as a property. If the property was assigned to some other model it would make sense to completely re-render the child.

I'm not convinced the distinction between state and ref is all that salient for child components. Assuming we're comparing passing the state value with passing ref.current, in either case it's just an instance being passed as a prop from the perspective of the child component. In either case it will be a stable reference (i.e. prevProp === thisProp) if the value hasn't changed, but prevProp !== thisProp when the value changes, so if it were used as a dependency in a useEffect (for instance), the right thing would happen. The distinction that matters is that with a ref, changing the value won't trigger a re-render of the parent component while with state it will. And, of course, absent use of `React.memo` or `PureComponent`, rendering the parent component will automatically render all of the child components in either case. So I think the distinction boils down to:

My convention would be to use a ref for values that aren't changed. If the need arose to change it in the future _and_ I wanted that change to cause a re-render I would change to using state at that point.

I take your convention to be: decide whether if this value were to change would we want it to re-render and use state from the outset in that case.

An argument for your convention is that it "bakes in" an understanding that the author of the code has at the time the code is written as a hint to a future author that needs to extend the code in a particular way (namely to set the value in question).

My counter-counter-counter-argument (😉 ) would be that it seems like premature generalization. We have no need to modify the value in question at the moment and no particular reason to expect to do so in the future. We should use the simplest implementation that meets the current need rather than build in an extension mechanism for which we have no particular expectation of its use. When I see state in React I assume it's there because it's a value that changes and should trigger a render when it does. From that perspective, this is an unconventional use of state. So the question is whether laying the groundwork for such future mutability is worth it in this context.

All of that said, we're still splitting hairs and I don't think it ultimately matters which implementation you choose, so do what makes sense to you.

Reply:
All good points. 

Reflecting on this more, I think what I'm doing is trying to find ways and reasons to avoid `useRef` when possible. It just looks confusing to me no matter how well I understand it.

So I'm going to leave this as is and I'll continue to struggle with it. 😄

Reply:
It was enjoyable to read. :wink: All the points make sense to me. 

I have no idea about the project and why the cloning happens here (and not somewhere higher where the copy can be properly kept (but I believe there's a reason)). However, looking at this tiny component, my first impression is that I'd rather use `useState`. It feels safer and I guess that's Scott's argument. If we use `useRef`, there's one more choice to be made:
1. Should we pass the whole ref to a child? =>`<ToolbarComponent  toolbarModel={toolbarModelRef}`
2. Should we pass the current value?  => `<ToolbarComponent toolbarModel={toolbarModelRef.current}`

Version 1. has its use cases, as it was mentioned - usually for some optimization. The child knows it's a ref and how to deal with that.

But version 2., which also was mentioned (I think), seems pretty error-prone. I think it'd work only with the assumption that this value naver changes. It's true here, but in general, the child might assume it's gonna re-render on the `toolbarModel` prop update, but it won't have a chance to do it, as the parent won't trigger re-render. But I think this was also covered by Kirk in the "premature generalization" argument. 😉